drivers: optee: Introduce basic OP-TEE driver to Zephyr

oleksiimoisieiev opened the pull request:
This introduces the TEE subsystem definition and OP-TEE driver basic implementation with the unit tests
This is preliminary implementation which implements tee syscalls except for shm_register and shm_unregister calls.
DID NOT TESTED on the real hardware, only unit tests.

dsemenets committed:
WA: subsys/fs: support block devices in shell

For flash device, littlefs should be located at the partition.
Current implementation of the fs shell is hardcoded for the flash device
and the compilation fails if STORAGE_PARTITION isn't defined in the
device tree.
Add options for block dev support. This allows to mount littlefs on the
block device from the shell, and also removes the compile time
dependency on STORAGE_PARTITION, if CONFIG_FS_LITTLEFS_BLK_DEV is set.

This is an incremental improvement, ideally, someone should implement
selection of the storage dev from the shell args.

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Mykyta Poturai <mykyta_poturai@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
drivers: tee: Add Tee driver generic API

Implementation of the generic API Trusted Execution Evironment driver.
This implements all basic syscalls, needed by tee driver, basic
defines and helpers to work with shared memory.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

oleksiimoisieiev committed:
tests: drivers: tee: Add basic tests for TEE driver

This introduces basic TEE driver unit tests, based on the dummy TEE
driver.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

oleksiimoisieiev committed:
drivers: tee: optee: Introducing OP-TEE driver implementation

This includes the basic TEE driver api implementation for OP-TEE.
It provides an interface to the OP-TEE TrustZone from the Normal space
to start sessions and request functions of the TA.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

oleksiimoisieiev committed:
drivers: tee: optee: Moving optee smc interface from optee client

Moved helper headers, needed to work with OP-TEE, which defines basic
defines, used during communication with Trust-Zone

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

Deedone committed:
drivers: lfs: Fix UB in lfs init

dev_id is used as a pointer at the end of the call chain, so passing it
as an int makes no sense and can cause crashes if the pointer and int
types have different sizes.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Deedone committed:
drivers: mmc: Init semaphore before use

irq_xref_fin is used in mmc_reset, which is called before the semaphore
is initialized. Move the semaphore initialization to an earlier point to
prevent crashes.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
dts: bindings: Introduce device-tree bindings for OP-TEE

This introduces dts mapping description for the optee node.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

oleksiimoisieiev committed:
tests: drivers: optee: Introduce unit tests for the OP-TEE driver

This should test OP-TEE driver basic functionality without
communication with real OPTEE OS.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

xakep-amatop commented:
you are adding a new driver and device abstraction it would be nice to add some documentation to zephyr/doc

xakep-amatop commented on tests/drivers/tee/tee/testcase.yaml:
```diff
@@ -0,0 +1,9 @@
+    timeout: 500
+    integration_platforms:
+      - native_posix_64
+    platform_allow: native_posix_64
```
should you add R-Car gen3 & gen4 boards ?

oleksiimoisieiev replied:
I think not. I see no reason of running those test on the target boards. The idea is that this are the unit tests, that doesn't have direct connection to the optee. For integration testing we have zephyr-optee-test app.
I'm preparing for MR to xen-troops so all fixes will include to the new MR.

oleksiimoisieiev commented:
> you are adding a new driver and device abstraction it would be nice to add some documentation to zephyr/doc

+1. I will add docs.

xakep-amatop committed:
drivers: tee: optee: add support of shared memory cmds

Add support of shared memory register/unregister to optee driver.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop commented on drivers/tee/optee/Kconfig:
```diff
@@ -0,0 +1,8 @@
+
+config OPTEE
+	bool "Enable optee driver in system"
+	depends on TEE
```
`depends on ARM || ARM64 || RISCV`

oleksiimoisieiev replied:
+

oleksiimoisieiev committed:
drivers: tee: Add TEE_CAPS, needed by optee_client to tee

optee_client should use tee driver interface header to get an access
to the common defines. Added missing definitions to the headerg

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

oleksiimoisieiev committed:
tests: drivers: optee: Update unit tests to support shm funcs

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

oleksiimoisieiev commented:
> you are adding a new driver and device abstraction it would be nice to add some documentation to zephyr/doc

Tee.h has doxygen notes across the file. 
I'm preparing for MR to xen-troops so all fixes will include to the new MR.

oleksiimoisieiev commented:
Closing this PR as the new one for zephyr on xen-troops should be created.
